File: solutions/clojure/057-word-search.clj
;; Problem 57: Word Search (LeetCode 79)
;; Difficulty: Med
;; Language: Clojure
;; 
(defn exist [board word]
  (let [rows (count board) cols (count (first board))
        dfs (fn dfs [r c i vis]
              (cond
                (= i (count word)) true
                (or (< r 0) (>= r rows) (< c 0) (>= c cols)) false
                (or (vis [r c]) (not= (get-in board [r c]) (nth word i))) false
                :else (let [vis (conj vis [r c])]
                        (some #(dfs (+ r (first %)) (+ c (second %)) (inc i) vis)
                              [[1 0] [-1 0] [0 1] [0 -1]]))))]
    (boolean (some (fn [r] (some (fn [c] (dfs r c 0 #{}))
                                 (range cols)))
                   (range rows)))))

File: solutions/typescript/057-word-search.ts
/**
 * Problem 57: Word Search (LeetCode 79)
 * Difficulty: Med
 * Language: TypeScript
 */
function exist(board: string[][], word: string): boolean {
  const rows = board.length, cols = board[0].length;
  const dfs = (r: number, c: number, i: number, visited: Set<string>): boolean => {
    if (i === word.length) return true;
    if (r < 0 || r >= rows || c < 0 || c >= cols) return false;
    const key = \`\${r},\${c}\`;
    if (visited.has(key) || board[r][c] !== word[i]) return false;
    const next = new Set([...visited, key]);
    return [[r+1,c],[r-1,c],[r,c+1],[r,c-1]].some(
      ([nr,nc]) => dfs(nr, nc, i+1, next)
    );
  };
  return board.some((row, r) =>
    row.some((_, c) => dfs(r, c, 0, new Set()))
  );
}
